apps/faucet/src/endpoints/faucet.service.ts

```diff
@@ -0,0 +1,221 @@
+import { BadRequestException, Injectable, Logger } from '@nestjs/common';
+import { Account, Address, Transaction, TransactionPayload } from '@multiversx/sdk-core';
+import { Mnemonic, UserSigner } from '@multiversx/sdk-wallet';
+import { FaucetSettings } from './entities/faucet.settings';
+import { NetworkConfig, ProxyNetworkProvider } from '@multiversx/sdk-network-providers/out';
+import qs from 'qs';
+import { readFileSync } from 'fs';
+import { UserSecretKey } from '@multiversx/sdk-wallet/out';
+import { ApiService } from '@multiversx/sdk-nestjs-http';
+import { CacheService } from '@multiversx/sdk-nestjs-cache';
+import { AppConfigService } from '../config/app-config.service';
+import { CacheInfo } from '@libs/common';
+import { Constants } from '@multiversx/sdk-nestjs-common';
+
+@Injectable()
+export class FaucetService {
+  private readonly logger: Logger;
+
+  private networkConfig?: NetworkConfig;
+
+  provider?: ProxyNetworkProvider;
+  faucetAccount?: Account;
+  signer?: UserSigner;
+
+  constructor(
+    private readonly cachingService: CacheService,
+    private readonly apiConfigService: AppConfigService,
+    private readonly apiService: ApiService,
+  ) {
+    this.logger = new Logger(FaucetService.name);
+
+    if (this.isFaucetEnabled()) {
+      this.provider = new ProxyNetworkProvider(this.apiConfigService.config.gatewayUrl);
+      this.faucetAccount = new Account(new Address(this.apiConfigService.config.faucetAddress));
+      const privateKeyMode = this.apiConfigService.config.faucetPrivateKeyMode;
+      let secretKey;
+      if (privateKeyMode === 'mnemonic') {
+        secretKey = Mnemonic.fromString(this.apiConfigService.config.faucetMnemonic).deriveKey();
+      } else { // pem file
+        const pemContent = readFileSync(this.apiConfigService.config.faucetPemPath);
+        secretKey = UserSecretKey.fromPem(pemContent.toString());
+      }
+      this.signer = new UserSigner(secretKey);
+    }
+  }
+
+  isFaucetEnabled(): boolean {
+    const faucetAddress = this.apiConfigService.config.faucetAddress;
+
+    return faucetAddress !== undefined && faucetAddress !== null && faucetAddress !== '';
+  }
+
+  shouldBypassCaptchaCheck(): boolean {
+    return this.apiConfigService.config.faucetRecaptchaBypass;
+  }
+
+  private async getNetworkConfig() {
+    if (!this.networkConfig) {
+      this.networkConfig = await this.provider?.getNetworkConfig();
+    }
+
+    return this.networkConfig;
+  }
+
+  async retrieveFunds(address: string, nonce: number | undefined = undefined, clientIp: string | undefined = undefined, captcha: string | undefined = undefined): Promise<boolean> {
```

Comment:
Instead of defaulting optional parameters (nonce, clientIp, captcha) to undefined, you can simply omit this since they are already implicitly undefined if not passed

Reply:
right. done